### ISSUE TITLE
Support CompactIndex API for FIPS

### ISSUE THREAD

**Issue by danepowell** [rubygems/rubygems#7684]
### Describe the problem as clearly as you can

The CompactIndex API uses MD5 which means it cannot be used on FIPS-enabled systems, resulting in this error message: `FIPS mode is enabled, bundler can't use the CompactIndex API`

Bundler then falls back to a full source index fetch which is prohibitively slow (timing out our CI jobs for instance).

If you check the "overriding fips" section here, it sounds like there should be ways for Ruby to flag the usage of md5 as safe since it's not being used for cryptographic purposes here.

Alternatively, could the CompactIndex API be converted to use a FIPS-compatible algorithm?

### Did you try upgrading rubygems & bundler?

Yes

### Post steps to reproduce the problem

Enable FIPS mode

### Which command did you run?

Run `bundle install`

### What were you expecting to happen?

Bundle install should use the CompactIndex API

### What actually happened?

Bundler falls back to source index fetching

**Comment by deivid-rodriguez** [comment]
I did some testing locally and it seems we may be able to temporary disable FIPS mode every time we need to with `OpenSSL.fips_mode = true/false`? Changing the API to serve checksums calculated with a different algorithm in a smooth way would be a bit of work...

**Comment by deivid-rodriguez** [comment]
> That sounds great. I'd rather be able to install gems without FIPS than not be able to install them at all.

The idea that I had would be to disable it temporary just for "bundler non crypto stuff", but then leave it enabled if it was enabled before. So you can still get the benefits of FIPS in other places.

**Comment by danepowell** [comment]
Right. Just to be clear: you're proposing a code change in Bundler to support temporarily disabling FIPS, right? Because I've tried at least a dozen ways of disabling FIPS (or at least allowing MD5) _outside_ of Bundler before running the install and AFAICT it's impossible.

**Comment by deivid-rodriguez** [comment]
Yes. Basically wrap our MD5 usages with something like

```
def with_fips_mode_disabled
  previous_mode = OpenSSL.fips_mode
  OpenSSL.fips_mode = false
  yield
ensure
  OpenSSL.fips_mode = false
end
```

and stop skipping the compact index API when MD5 is not available.
